app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java

```diff
@@ -191,4 +202,145 @@ public void onNewDataReceivedEvent(final NewDataReceivedEvent event) {
         }
     }
 
+    /**
+     * Denotes the cluster to which a certain tap interval belongs
+     */
+    private enum Cluster {
+        SHORT,
+        LONG
+    }
+
+    /**
+     * Executes K-means++ clustering on the intervals to divide them into a short and long cluster
+     */
+    public static final class KMeansClusterHelper {
+        private List<Long> intervals;
+        private List<Long> clusterCenter1;
+        private List<Long> clusterCenter2;
+
+        private KMeansClusterHelper() {
+            // Prevent instantiation from outside the package
+        }
+
+        private static List<Long> getIntervals(final List<SingleTap> taps) {
+            final List<Long> intervals = new ArrayList<>(taps.size() - 1);
+            for (int i = 0; i < taps.size() - 1; i++) {
+                intervals.add(taps.get(i + 1).getTimestamp().getTime() - taps.get(i).getTimestamp().getTime());
+            }
+            return intervals;
+        }
+
+        /**
+         * Returns the unique key that corresponds to the taps specified.
+         * The key is a binary number.
+         * Short interval = 0, long interval = 1
+         * If the user taps a short - long - short - long interval, then the returned key is
+         * 1010 (because the numbers are always written from right to left
+         *
+         * @param taps The taps that should be encoded to a key
+         * @return The unique key that corresponds to the taps
+         */
+        int getEncryptionKey(final List<SingleTap> taps) {
+            this.clusterCenter1 = new ArrayList<>(taps.size());
+            this.clusterCenter2 = new ArrayList<>(taps.size());
+            this.intervals = getIntervals(taps);
+            long clusterCenterMean1 = this.makeAndReturnFirstClusterMean();
+            long clusterCenterMean2 = this.makeAndReturnSecondClusterMean();
+
+            while (!this.intervals.isEmpty()) {
+                this.addClosestTimestampToCluster(clusterCenterMean1, clusterCenterMean2);
+                final ImmutablePair<Long, Long> newClusterCenters = this.recalculateClusterCenters();
+                clusterCenterMean1 = newClusterCenters.getLeft();
+                clusterCenterMean2 = newClusterCenters.getRight();
+            }
+
+            final long breakpoint = this.getBreakpoint();
+            this.intervals = getIntervals(taps);
+            return this.generateKey(breakpoint);
+        }
+
+        private long makeAndReturnFirstClusterMean() {
```

Comment:
Here I agree with you, because these the purpose of the methods is not very clear indeed